### PR TITLE
Add Not to allow predicate negation

### PR DESCRIPTION
**Pull request by adriangb**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -128,15 +128,23 @@ Users should prefer the statically inspectable metadata above, but if you need
...
+* `IsNotNan = Predicate(Not(math.isnan))`
+* `IsInfinite = Predicate(math.isinf)`
+* `IsNotInfinite = Predicate(Not(math.isinf))`
+
 Some libraries might have special logic to handle known or understandable predicates,
 for example by checking for `str.isdigit` and using its presence to both call custom
 logic to enforce digit-only strings, and customise some generated external schema.
 Users are therefore encouraged to avoid indirection like `lambda s: s.lower()`, in
 favor of introspectable methods such as `str.lower` or `re.compile("pattern").search`.
+To enable basic negation of commonly used predicates like `math.isnan` without introducing introspection that makes it impossible for implementers to introspect the predicate we provide a `Not` wrapper that simply negates the predicate in an introspectable manner. Several of the predicates listed above are created in this manner.
```
Comment (Zac-HD) [inline README.md:147]: Break line for consistency?

**Review by Zac-HD — APPROVED** [review]
Seems reasonable; it won't actually help Hypothesis but nor will it hurt and it's simple to explain and implement.

- Consider [defining `Predicate.__invert__`](https://docs.python.org/3/reference/datamodel.html#object.__invert__) so we can write `~Predicate[fn]` as sugar for `Predicate[Not[fn]]`?
- I'd favor runtime checks (`if __debug__:`) that (a) whatever `Not[...]` wraps is a callable and not an instance of `BaseMetadata`, and (b) that whatever non-`Predicate` `BaseMetadata` classes wrap is _not_ a `Not`.  Seems kinda silly, but it's the kind of thing that can avoid a surprising amount of pain later.  Perf impact seems small, but maybe we really care about it anyway 🤔

**Comment by adriangb** [comment]
> I'd favor runtime checks (if __debug__:) that (a) whatever Not[...] wraps is a callable and not an instance of BaseMetadata, and (b) that whatever non-Predicate BaseMetadata classes wrap is not a Not. Seems kinda silly, but it's the kind of thing that can avoid a surprising amount of pain later. Perf impact seems small, but maybe we really care about it anyway 🤔

Do you mean that we check to prohibit `Not[Predicate[...]]` and only allow `Predicate[Not[...]]`?

**Comment by Zac-HD** [comment]
Either of those seem fine, but I'd ban eg `Not[Ge(5)]` and `MinLe([Not[3])`.

**Comment by adriangb** [comment]
> MinLen([Not[3])

Won't this already be rejected by type checkers? `Not` requires a callable and `3` is not a callable. Similarly `MinLen` requires an `int` as the first argument and `Not` is not an`int`.

**Comment by Zac-HD** [comment]
Yeah, seems fine to rely on static type checking if we expect users to have that.

**Comment by adriangb** [comment]
I think that’s a fair bet and I’d rather rely on that than adding more runtime type checking here.

File: README.md (hunk 1 above)
Comment (adriangb) [reply README.md:147]:
```suggestion

To enable basic negation of commonly used predicates like `math.isnan` without introducing introspection that makes it impossible for implementers to introspect the predicate we provide a `Not` wrapper that simply negates the predicate in an introspectable manner. Several of the predicates listed above are created in this manner.
```

**Pull request merged**